Token Print Wizard

davidbernard opened the pull request:
Added :layout to Wizard Components
Wizard components with layouts are used to specify the layout of tokens.
Added PDFLayoutEngine to generate PDF of tokens from Wizard components.
Added :token_layout to ServiceProvider.
In seeds.db a token layout is added to the registration service provider and is the default for token generation.
Updated cucumber tests and unit tests.

davidbernard committed:
Added :layout to Wizard Components
Wizard components with layouts are used to specify the layout of tokens.
Added PDFLayoutEngine to generate PDF of tokens from Wizard components.
Added :token_layout to ServiceProvider.
In seeds.db a token layout is added to the registration service provider and is the default for token generation.
Updated cucumber tests and unit tests.

athurn commented on app/controllers/people_controller.rb:
```diff
@@ -16,7 +16,10 @@ def show
       }
       format.pdf {
-          pdf = Services::Token.generate_token_pdf(@person, settings.token_media)
+          Rails.logger.info "settings #{settings.token_media}"
```
settings?

davidbernard replied:
application_controler.rb publishes Rails.application.config as settings. token_media is currently a global settings but could be moved into service_provider

athurn commented on app/models/service_provider.rb:
```diff
@@ -9,6 +9,7 @@ class ServiceProvider
     many :person, :in => :clients
 
     one :wizard, class: Wizard::Wizard
+    one :token_layout, class: Wizard::Wizard
```
I am not sure, if the printing layout is actually a Wizard :)

athurn commented on app/services/pdf_layout_engine.rb:
```diff
@@ -0,0 +1,44 @@
+                    layout(object, inner_component, document, w, h, l, t )
+                end
+            else
+                if component.template_name == 'wizard/components/qr' then
```
it would be cool, if the component could take care of rendering itself rather than having some if elses here.

davidbernard replied:
I thought about this. Components currently are information structures linked to presentation through template name derived from the class. I didn't want to couple essentially an information structure with layout functionality given that different layout engines could have different requirements. So the implementation is simplistic until we come up with a good scheme.

athurn commented on app/services/token.rb:
```diff
@@ -27,28 +27,15 @@ def self.generate_token_pdf(person, media)
         parameters = media_parameters[media]
-        Rails.logger.info "Generating token for #{person.id} #{media}"
+        Rails.logger.info "Parameters #{parameters}"
         width = parameters[:width].send(:mm)
```
cannot we just to parameters[:width].mm?

davidbernard replied:
fixed.

athurn commented on features/token.feature:
```diff
@@ -0,0 +1,13 @@
+| input2-2  | Input 2 - 2 |
+| input3-1  | Input 3 - 1 |
+| input3-2  | Input 3 - 2 |
+Then I should see a token
```
theres a step missing here - you don't just see a token because of an entry the database :) a cukes is normally given, when, then

davidbernard replied:
capybara doesn't handle file download and the people_controller generates the token on the fly. Changed to "Then I can generate and see a token". There may be a use case to change the floor of people_controller. Can update the test then.

athurn commented on app/controllers/people_controller.rb:
```diff
@@ -16,7 +16,10 @@ def show
+          Rails.logger.info "settings #{settings.token_media}"
+          
+          pdf = Services::Token.generate_token_pdf(@person, settings.token_media,
+                                        ServiceProvider.find_by_special_role(:registration).token_layout)
```
we could go crazy and allow to generate tokens for each service provider, I am not sure, but maybe we should get rid of an service provider agnostic view.
woudnt it be cool if we could generate a pdf somehow like this: /service_providers/registration/users/ID123.pdf

davidbernard replied:
Agreed. The print wizard stuff could be used to generate specific paper forms for the service provider. Alternatively we could try using a report generator package. I am reasonable comfortable with the ability to layout at least simple single page forms using the current structure. These user stories need to be developed further.

athurn commented on app/models/print_layout/layout.rb:
```diff
@@ -0,0 +1,11 @@
+        include ActiveModel::Model
+        include MongoMapper::EmbeddedDocument
+    
+        key :height, Float
```
does this need to be a float?

davidbernard replied:
Using proportions to enclosing components for height/width and position.

athurn commented on app/models/wizard/component.rb:
```diff
@@ -2,7 +2,9 @@ module Wizard
     include MongoMapper::EmbeddedDocument
-
+    
+    one :layout, class: PrintLayout::Layout
```
components are currently layout agnostic, they just know how to draw themselves but not how to layout.

athurn commented on app/services/token.rb:
```diff
@@ -27,28 +27,15 @@ def self.generate_token_pdf(person, media)
            'A4' =>  { :width => 90, :height =>29, :leftMargin => 2, :rightMargin => 3 , :spacing => 3 } }
```
we might actually want to move this somewhere else, too. shouldn't the printing layout specify which paper to use? but that's fine for the moment :)

athurn commented:
awesome!! :) First pull request this year :). I need to start working on my card otherwise you'll outdistance me.
I'd a couple of comments, you can see them inline

davidbernard committed:
Minor fixes in response to review comments and to ensure test pass

Disable printer service test as environment not setup in CI
Change name of step in token features.
Removed unnecessary "send()"

StephenSmithwick commented:
Is this pull request ready to be merged?

davidbernard commented:
Sorry for the delay. Been overseas.

The pull request can be merged but github is reporting merge conflicts.

I don’t have time at the moment to sort it out but please have a go.


> On 14 May 2015, at 2:47 pm, Stephen Smithwick <notifications@github.com> wrote:
> 
> Is this pull request ready to be merged?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/FirstStop/DisasterRegistration/pull/19#issuecomment-101908254>.
>
